src/Test/Request/TestClient.php

```diff
@@ -0,0 +1,186 @@
+    public function getHandlers(): array
+    {
+        return $this->handlers;
+    }
+
+    /**
+     * @param Handler $closure
+     */
+    public function addHandler(\Closure $closure, string|int|null $name = null): void
+    {
+        if ($name !== null) {
+            $this->handlers[$name] = $closure;
+        } else {
+            $this->handlers[] = $closure;
+        }
+    }
+
+    public function removeHandler(string|int $name): void
+    {
+        unset($this->handlers[$name]);
+    }
+
+    /**
+     * @throws ApiException|\JsonException
+     */
+    public function request(TestRequestContext $context): ResponseInterface
+    {
+        $response = null;
+
+        foreach ($this->getHandlers() as $handler) {
+            if ($handlerResponse = $handler($context)) {
+                $response = $this->handlerResponseToResponse($handlerResponse, $context);
+
+                break;
+            }
+        }
+
+        if (!$response) {
+            throw new ApiException(
+                ApiException::CODE_INVALID_REQUEST,
+                'No response given',
+                $this->factory->createResponse(400),
+            );
+        }
+
+        $this->requests[] = TestResponseContext::from($context, $response);
+
+        return $response;
+    }
+
+    /**
+     * Resets all logged requests.
+     */
+    public function reset(): void
+    {
+        $this->requests = [];
+    }
+
+    /**
+     * @return list<TestResponseContext>
+     */
+    public function getAll(): array
+    {
+        return $this->requests;
+    }
+
+    public function get(int $idx): ?TestResponseContext
+    {
+        return $this->requests[$idx] ?? null;
+    }
+
+    public function getLast(): ?TestResponseContext
+    {
+        return \end($this->requests) ?: null;
+    }
+
+    public function getFirst(): ?TestResponseContext
+    {
+        return $this->requests[0] ?? null;
+    }
+
+    /**
+     * @param Filter $closure
+     */
+    public function firstWhere(\Closure $closure): ?TestResponseContext
+    {
+        return \current(\array_filter($this->requests, $closure, \ARRAY_FILTER_USE_BOTH)) ?: null;
+    }
+
+    /**
+     * @param Filter $closure
+     */
+    public function lastWhere(\Closure $closure): ?TestResponseContext
+    {
+        return \current(\array_filter(\array_reverse($this->requests, true), $closure, \ARRAY_FILTER_USE_BOTH)) ?: null;
+    }
+
+    /**
+     * @param Response $response
+     */
+    protected function handlerResponseToResponse(mixed $response, TestRequestContext $context): ResponseInterface
+    {
+        if ($response instanceof ResponseInterface) {
+            return $response;
+        }
+
+        if (!\is_array($response)) {
+            return $this->handlerResponseToResponse([200, $response], $context);
+        }
+
+        if (!\array_is_list($response) || \count($response) !== 2) {
+            return $this->handlerResponseToResponse([200, $response], $context);
+        }
```

Comment:
Is there a reason why you aren't combining them?

```suggestion
        if (!\is_array($response) || !\array_is_list($response) || \count($response) !== 2) {
            return $this->handlerResponseToResponse([200, $response], $context);
        }
```